File: backend/app/services/receipt_service.py
import io
from datetime import datetime
from typing import Any, Dict
from fastapi import HTTPException, status
from reportlab.lib.pagesizes import A5, letter
from reportlab.lib import colors
from reportlab.platypus import SimpleDocTemplate, Paragraph, Spacer, Table, TableStyle, HRFlowable
from reportlab.lib.styles import getSampleStyleSheet, ParagraphStyle
from app.core.config import settings
from app.services.transaction_service import transaction_service
from app.utils.currency import format_rupiah
from app.utils.formatting import format_datetime, format_kg


class ReceiptService:
    @classmethod
    def get_receipt_data(cls, transaction_id: str, current_user: Dict[str, Any]) -> Dict[str, Any]:
        """Fetch and format receipt data for viewing or printing."""
        tx = transaction_service.get_transaction_by_id(transaction_id)
        
        # Access control: If user is NASABAH, verify ownership
        if current_user.get("role") == "NASABAH":
            if current_user.get("nasabah_id") != tx["nasabah_id"]:
                raise HTTPException(
                    status_code=status.HTTP_403_FORBIDDEN,
                    detail="Anda tidak memiliki izin untuk melihat bukti transaksi nasabah lain.",
                )

        now_str = datetime.now().strftime("%d/%m/%Y %H:%M")
        
        detail_data = None
        if tx["type"] == "SETOR":
            detail_data = {
                "category_code": tx["category"].get("code") if tx.get("category") else None,
                "category_name": tx["category"]["name"] if tx.get("category") else None,
                "weight_kg": tx["weight_kg"],
                "weight_formatted": format_kg(tx["weight_gram"]),
                "price_per_kg": tx["price_per_kg"],
                "price_formatted": f"{format_rupiah(tx['price_per_kg'])}/kg",
                "amount": tx["amount"],
                "amount_formatted": format_rupiah(tx["amount"]),
            }
        else:
            detail_data = {
                "category_code": None,
                "category_name": None,
                "weight_kg": None,
                "weight_formatted": None,
                "price_per_kg": None,
                "price_formatted": None,
                "amount": tx["amount"],
                "amount_formatted": format_rupiah(tx["amount"]),
            }

        return {
            "app_name": settings.BANK_NAME,
            "title": "BUKTI TRANSAKSI TABUNGAN BANK SAMPAH",
            "transaction_no": tx["transaction_no"],
            "transaction_date": tx["transaction_date"],
            "transaction_date_formatted": format_datetime(tx["transaction_date"]),
            "type": tx["type"],
            "type_display": "SETOR SAMPAH" if tx["type"] == "SETOR" else "TARIK TUNAI",
            "nasabah": {
                "id": tx["nasabah_id"],
                "customer_id": tx["nasabah_customer_id"],
                "account_no": tx["nasabah_customer_id"],
                "name": tx["nasabah_name"],
                "nik": tx["nasabah_nik"],
                "phone": tx.get("nasabah_phone"),
                "address": tx.get("nasabah_address"),
            },
            "detail": detail_data,
            "mutation": {
                "debit": tx["debit"],
                "debit_formatted": format_rupiah(tx["debit"]),
                "credit": tx["credit"],
                "credit_formatted": format_rupiah(tx["credit"]),
                "balance_after": tx["balance_after"],
                "balance_after_formatted": format_rupiah(tx["balance_after"]),
            },
            "notes": tx.get("notes") or "-",
            "footer": settings.RECEIPT_FOOTER,
            "printed_at": now_str,
        }

    @classmethod
    def generate_receipt_pdf(cls, transaction_id: str, current_user: Dict[str, Any]) -> bytes:
        """Generate PDF binary stream for a transaction receipt using ReportLab."""
        receipt = cls.get_receipt_data(transaction_id, current_user)
        
        buffer = io.BytesIO()
        doc = SimpleDocTemplate(
            buffer,
            pagesize=A5,
            rightMargin=24,
            leftMargin=24,
            topMargin=24,
            bottomMargin=24
        )
        
        styles = getSampleStyleSheet()
        title_style = ParagraphStyle(
            "TitleStyle",
            parent=styles["Heading2"],
            fontSize=13,
            leading=16,
            textColor=colors.HexColor("#15803d"),
            alignment=1, # Center
            spaceAfter=2
        )
        subtitle_style = ParagraphStyle(
            "SubTitleStyle",
            parent=styles["Normal"],
            fontSize=9,
            leading=12,
            textColor=colors.HexColor("#4b5563"),
            alignment=1,
            spaceAfter=10
        )
        section_heading = ParagraphStyle(
            "SectionHeading",
            parent=styles["Heading4"],
            fontSize=10,
            leading=12,
            textColor=colors.HexColor("#1f2937"),
            spaceBefore=6,
            spaceAfter=4
        )
        normal_style = ParagraphStyle(
            "ReceiptNormal",
            parent=styles["Normal"],
            fontSize=8.5,
            leading=11,
            textColor=colors.HexColor("#111827")
        )
        normal_right = ParagraphStyle(
            "ReceiptNormalRight",
            parent=normal_style,
            alignment=2
        )
        bold_style = ParagraphStyle(
            "ReceiptBold",
            parent=normal_style,
            fontName="Helvetica-Bold"
        )
        bold_right = ParagraphStyle(
            "ReceiptBoldRight",
            parent=normal_right,
            fontName="Helvetica-Bold"
        )
        th_left = ParagraphStyle(
            "ReceiptTHL",
            parent=bold_style,
            textColor=colors.HexColor("#166534"),
            alignment=0
        )
        th_right = ParagraphStyle(
            "ReceiptTHR",
            parent=bold_right,
            textColor=colors.HexColor("#166534"),
            alignment=2
        )
        footer_style = ParagraphStyle(
            "ReceiptFooter",
            parent=styles["Normal"],
            fontSize=8,
            leading=10,
            textColor=colors.HexColor("#6b7280"),
            alignment=1
        )

        elements = []
        
        # Header
        elements.append(Paragraph(receipt["app_name"].upper(), title_style))
        elements.append(Paragraph(receipt["title"], subtitle_style))
        elements.append(HRFlowable(width="100%", thickness=1, color=colors.HexColor("#d1d5db"), spaceAfter=8))

        # Metadata Table
        meta_data = [
            [Paragraph("<b>No. Transaksi:</b>", normal_style), Paragraph(receipt["transaction_no"], bold_style),
             Paragraph("<b>Tanggal:</b>", normal_style), Paragraph(receipt["transaction_date_formatted"], normal_style)],
            [Paragraph("<b>ID Nasabah:</b>", normal_style), Paragraph(receipt["nasabah"]["customer_id"], bold_style),
             Paragraph("<b>NIK:</b>", normal_style), Paragraph(receipt["nasabah"]["nik"], normal_style)],
            [Paragraph("<b>Nama Nasabah:</b>", normal_style), Paragraph(receipt["nasabah"]["name"], normal_style),
             Paragraph("<b>Jenis:</b>", normal_style), Paragraph(receipt["type_display"], bold_style)],
        ]
        meta_table = Table(meta_data, colWidths=[80, 110, 55, 115])
        meta_table.setStyle(TableStyle([
            ('VALIGN', (0, 0), (-1, -1), 'TOP'),
            ('BOTTOMPADDING', (0, 0), (-1, -1), 3),
            ('TOPPADDING', (0, 0), (-1, -1), 3),
            ('LEFTPADDING', (0, 0), (-1, -1), 2),
            ('RIGHTPADDING', (0, 0), (-1, -1), 2),
        ]))
        elements.append(meta_table)
        elements.append(Spacer(1, 8))

        # Details Table
        if receipt["type"] == "SETOR" and receipt["detail"]:
            elements.append(Paragraph("Detail Setoran", section_heading))
            det = receipt["detail"]
            detail_data = [
                [
                    Paragraph("<b>Kelompok Sampah</b>", th_left),
                    Paragraph("<b>Berat</b>", th_right),
                    Paragraph("<b>Harga/kg</b>", th_right),
                    Paragraph("<b>Total Nilai</b>", th_right)
                ],
                [
                    Paragraph(det["category_name"] or "-", normal_style),
                    Paragraph(det["weight_formatted"] or "-", normal_right),
                    Paragraph(det["price_formatted"] or "-", normal_right),
                    Paragraph(det["amount_formatted"], normal_right)
                ]
            ]
            detail_table = Table(detail_data, colWidths=[140, 70, 75, 75])
            detail_table.setStyle(TableStyle([
                ('BACKGROUND', (0, 0), (-1, 0), colors.HexColor("#f0fdf4")),
                ('GRID', (0, 0), (-1, -1), 0.5, colors.HexColor("#e5e7eb")),
                ('VALIGN', (0, 0), (-1, -1), 'TOP'),
                ('TOPPADDING', (0, 0), (-1, -1), 4),
                ('BOTTOMPADDING', (0, 0), (-1, -1), 4),
                ('LEFTPADDING', (0, 0), (-1, -1), 4),
                ('RIGHTPADDING', (0, 0), (-1, -1), 4),
            ]))
            elements.append(detail_table)
            elements.append(Spacer(1, 8))

        # Mutation & Balance Table
        elements.append(Paragraph("Mutasi & Saldo Tabungan", section_heading))
        mut = receipt["mutation"]
        mutation_data = [
            [Paragraph("Debit (Penarikan)", normal_style), Paragraph(mut["debit_formatted"], normal_right)],
            [Paragraph("Kredit (Setoran)", normal_style), Paragraph(mut["credit_formatted"], normal_right)],
            [Paragraph("<b>Saldo Akhir Tabungan</b>", bold_style), Paragraph(f"<b>{mut['balance_after_formatted']}</b>", bold_right)],
        ]
        mutation_table = Table(mutation_data, colWidths=[230, 130])
        mutation_table.setStyle(TableStyle([
            ('GRID', (0, 0), (-1, -1), 0.5, colors.HexColor("#e5e7eb")),
            ('BACKGROUND', (0, 2), (-1, 2), colors.HexColor("#f9fafb")),
            ('VALIGN', (0, 0), (-1, -1), 'TOP'),
            ('TOPPADDING', (0, 0), (-1, -1), 4),
            ('BOTTOMPADDING', (0, 0), (-1, -1), 4),
            ('LEFTPADDING', (0, 0), (-1, -1), 4),
            ('RIGHTPADDING', (0, 0), (-1, -1), 4),
        ]))
        elements.append(mutation_table)
        elements.append(Spacer(1, 8))

        if receipt.get("notes") and receipt["notes"] != "-":
            elements.append(Paragraph(f"<b>Catatan:</b> {receipt['notes']}", normal_style))
            elements.append(Spacer(1, 8))

        # Footer
        elements.append(HRFlowable(width="100%", thickness=0.5, color=colors.HexColor("#e5e7eb"), spaceAfter=6))
        elements.append(Paragraph(receipt["footer"], footer_style))
        elements.append(Paragraph(f"Dicetak pada: {receipt['printed_at']}", footer_style))

        doc.build(elements)
        buffer.seek(0)
        return buffer.getvalue()


receipt_service = ReceiptService()

File: backend/src/services/receiptService.ts
import PDFDocRaw from 'pdfkit/js/pdfkit.standalone.js';
import { config } from '../core/config.js';
import { AppError } from '../utils/response.js';
import { formatRupiah } from '../utils/currency.js';
import { formatDateTime, formatKg } from '../utils/formatting.js';
import { transactionService } from './transactionService.js';

// Standalone PDFKit with inlined standard AFM fonts for Cloudflare Workers & Node.js edge environments
const PDFDocument = ((PDFDocRaw as any)?.default || PDFDocRaw) as typeof import('pdfkit');

export class ReceiptService {
  public static async getReceiptData(transactionId: string, currentUser: any) {
    const tx = await transactionService.getTransactionById(transactionId);

    // Access control: If user is NASABAH, verify ownership
    if (currentUser.role === 'NASABAH') {
      if (currentUser.nasabah_id !== tx.nasabah_id) {
        throw new AppError(
          'Anda tidak memiliki izin untuk melihat bukti transaksi nasabah lain.',
          403,
          'FORBIDDEN'
        );
      }
    }

    const now = new Date();
    const nowStr = formatDateTime(now);

    let detailData = null;
    if (tx.type === 'SETOR') {
      detailData = {
        category_code: tx.category ? tx.category.price_code : null,
        category_name: tx.category ? tx.category.name : null,
        weight_kg: tx.weight_kg,
        weight_formatted: formatKg(tx.weight_gram),
        price_per_kg: tx.price_per_kg,
        price_formatted: `${formatRupiah(tx.price_per_kg)}/kg`,
        amount: tx.amount,
        amount_formatted: formatRupiah(tx.amount),
      };
    } else {
      detailData = {
        category_code: null,
        category_name: null,
        weight_kg: null,
        weight_formatted: null,
        price_per_kg: null,
        price_formatted: null,
        amount: tx.amount,
        amount_formatted: formatRupiah(tx.amount),
      };
    }

    return {
      app_name: config.BANK_NAME,
      title: 'BUKTI TRANSAKSI TABUNGAN BANK SAMPAH',
      transaction_no: tx.transaction_no,
      transaction_date: tx.transaction_date,
      transaction_date_formatted: formatDateTime(tx.transaction_date),
      type: tx.type,
      type_display: tx.type === 'SETOR' ? 'SETOR SAMPAH' : 'TARIK TUNAI',
      officer_name: tx.officer_name || (currentUser?.role === 'ADMIN' ? (currentUser.name || currentUser.username) : 'Admin BSU'),
      nasabah: {
        id: tx.nasabah_id,
        customer_id: tx.nasabah_customer_id,
        account_no: tx.nasabah_customer_id,
        name: tx.nasabah_name,
        nik: tx.nasabah_nik,
        phone: tx.nasabah_phone,
        address: tx.nasabah_address,
      },
      detail: detailData,
      items: (tx as any).items || [],
      mutation: {
        debit: tx.debit,
        debit_formatted: formatRupiah(tx.debit),
        credit: tx.credit,
        credit_formatted: formatRupiah(tx.credit),
        balance_after: tx.balance_after,
        balance_after_formatted: formatRupiah(tx.balance_after),
      },
      notes: tx.notes || '-',
      footer: config.RECEIPT_FOOTER,
      printed_at: nowStr,
    };
  }

  public static async generateReceiptPdf(
    transactionId: string,
    currentUser: any
  ): Promise<Buffer> {
    const receipt = await this.getReceiptData(transactionId, currentUser);

    return new Promise((resolve, reject) => {
      try {
        const doc = new PDFDocument({
          size: 'A5',
          margin: 28,
        });

        const chunks: Buffer[] = [];
        doc.on('data', (chunk: any) => {
          chunks.push(Buffer.isBuffer(chunk) ? chunk : Buffer.from(chunk));
        });
        doc.on('end', () => resolve(Buffer.concat(chunks)));
        doc.on('error', (err: any) => reject(err));

      // Header
      doc
        .fontSize(14)
        .font('Helvetica-Bold')
        .fillColor('#15803d')
        .text(receipt.app_name.toUpperCase(), { align: 'center' });

      doc
        .fontSize(10)
        .font('Helvetica-Bold')
        .fillColor('#1f2937')
        .text(receipt.title, { align: 'center' });

      doc.moveDown(0.3);
      doc
        .fontSize(8)
        .font('Helvetica')
        .fillColor('#6b7280')
        .text(`No. Transaksi: ${receipt.transaction_no}  |  Tanggal: ${receipt.transaction_date_formatted}  |  Petugas: ${receipt.officer_name || 'Petugas Loket'}`, {
          align: 'center',
        });

      doc.moveDown(0.5);
      doc
        .strokeColor('#e5e7eb')
        .lineWidth(1)
        .moveTo(28, doc.y)
        .lineTo(doc.page.width - 28, doc.y)
        .stroke();
      doc.moveDown(0.6);

      // Section: Data Nasabah
      doc
        .fontSize(9)
        .font('Helvetica-Bold')
        .fillColor('#111827')
        .text('DATA NASABAH');
      doc.moveDown(0.3);

      const drawRow = (label: string, value: string) => {
        const y = doc.y;
        doc
          .fontSize(8)
          .font('Helvetica')
          .fillColor('#4b5563')
          .text(label, 32, y, { width: 110 });
        doc
          .fontSize(8)
          .font('Helvetica-Bold')
          .fillColor('#1f2937')
          .text(`:  ${value || '-'}`, 145, y, {
            width: doc.page.width - 145 - 32,
          });
        doc.moveDown(0.25);
      };

      drawRow('ID Nasabah / Rek', receipt.nasabah.customer_id);
      drawRow('Nama Nasabah', receipt.nasabah.name);
      drawRow('No. KTP / NIK', receipt.nasabah.nik);
      if (receipt.nasabah.phone) drawRow('No. HP', receipt.nasabah.phone);
      if (receipt.nasabah.address) drawRow('Alamat', receipt.nasabah.address);

      doc.moveDown(0.5);
      doc
        .strokeColor('#e5e7eb')
        .lineWidth(1)
        .moveTo(28, doc.y)
        .lineTo(doc.page.width - 28, doc.y)
        .stroke();
      doc.moveDown(0.6);

      // Section: Detail Transaksi
      doc
        .fontSize(9)
        .font('Helvetica-Bold')
        .fillColor('#111827')
        .text(`DETAIL TRANSAKSI (${receipt.type_display})`);
      doc.moveDown(0.3);

      if (receipt.type === 'SETOR' && receipt.items && receipt.items.length > 0) {
        doc.fontSize(8).font('Helvetica-Bold').fillColor('#374151');
        const startY = doc.y;
        doc.text('NO', 32, startY, { width: 20 });
        doc.text('JENIS / KELOMPOK SAMPAH', 55, startY, { width: 145 });
        doc.text('BERAT', 205, startY, { width: 50, align: 'right' });
        doc.text('TARIF', 260, startY, { width: 60, align: 'right' });
        doc.text('SUBTOTAL', 325, startY, { width: doc.page.width - 325 - 32, align: 'right' });
        doc.moveDown(0.3);

        doc.strokeColor('#d1d5db').lineWidth(0.5).moveTo(32, doc.y).lineTo(doc.page.width - 32, doc.y).stroke();
        doc.moveDown(0.3);

        receipt.items.forEach((it: any, idx: number) => {
          const rowY = doc.y;
          doc.fontSize(8).font('Helvetica');
          const nameText = it.display_name || it.category_name || '-';
          const nameHeight = doc.heightOfString(nameText, { width: 145 });
          const rowHeight = Math.max(nameHeight, 12);

          doc.fillColor('#4b5563');
          doc.text(String(idx + 1), 32, rowY, { width: 20 });
          doc.text(nameText, 55, rowY, { width: 145 });
          doc.text(formatKg(it.weight_gram), 205, rowY, { width: 50, align: 'right' });
          doc.text(formatRupiah(it.price_per_kg), 260, rowY, { width: 60, align: 'right' });
          doc.font('Helvetica-Bold').fillColor('#15803d');
          doc.text(formatRupiah(it.amount), 325, rowY, { width: doc.page.width - 325 - 32, align: 'right' });

          doc.y = rowY + rowHeight + 4;
        });

        doc.strokeColor('#d1d5db').lineWidth(0.5).moveTo(32, doc.y).lineTo(doc.page.width - 32, doc.y).stroke();
        doc.moveDown(0.3);

        const totalY = doc.y;
        doc.fontSize(8).font('Helvetica-Bold').fillColor('#111827');
        doc.text('TOTAL:', 55, totalY, { width: 145 });
        doc.text(receipt.detail?.weight_formatted || '-', 205, totalY, { width: 50, align: 'right' });
        doc.text(receipt.detail?.amount_formatted || '-', 325, totalY, { width: doc.page.width - 325 - 32, align: 'right' });
        doc.moveDown(0.4);
      } else if (receipt.type === 'SETOR' && receipt.detail) {
        drawRow('Kelompok Sampah', receipt.detail.category_name || '-');
        drawRow('Berat Sampah', receipt.detail.weight_formatted || '-');
        drawRow('Harga Satuan', receipt.detail.price_formatted || '-');
        drawRow('Total Setoran', receipt.detail.amount_formatted || '-');
      } else {
        drawRow('Jenis Penarikan', 'Tarik Tunai Saldo');
        drawRow('Nominal Tarik', receipt.mutation.debit_formatted);
      }

      if (receipt.notes && receipt.notes !== '-') {
        drawRow('Catatan', receipt.notes);
      }

      doc.moveDown(0.5);
      doc
        .strokeColor('#e5e7eb')
        .lineWidth(1)
        .moveTo(28, doc.y)
        .lineTo(doc.page.width - 28, doc.y)
        .stroke();
      doc.moveDown(0.6);

      // Section: Mutasi Saldo Box
      const boxY = doc.y;
      doc
        .rect(28, boxY, doc.page.width - 56, 42)
        .fillAndStroke('#f0fdf4', '#86efac');

      doc
        .fontSize(9)
        .font('Helvetica')
        .fillColor('#166534')
        .text('Saldo Akhir Tabungan:', 40, boxY + 8);

      doc
        .fontSize(14)
        .font('Helvetica-Bold')
        .fillColor('#15803d')
        .text(receipt.mutation.balance_after_formatted, 40, boxY + 20);

      doc.y = boxY + 52;
      doc.moveDown(0.5);

      // Section: Tanda Tangan
      const sigY = doc.y;
      const colWidth = (doc.page.width - 56) / 2;

      doc.fontSize(8).font('Helvetica').fillColor('#374151');
      doc.text('Nasabah,', 28, sigY, { width: colWidth, align: 'center' });
      doc.text('Petugas Bank Sampah,', 28 + colWidth, sigY, { width: colWidth, align: 'center' });

      const sigLineY = sigY + 36;
      doc.strokeColor('#d1d5db').lineWidth(0.5);
      doc.moveTo(48, sigLineY).lineTo(48 + colWidth - 40, sigLineY).stroke();
      doc.moveTo(28 + colWidth + 20, sigLineY).lineTo(doc.page.width - 48, sigLineY).stroke();

      doc.fontSize(8).font('Helvetica-Bold').fillColor('#111827');
      doc.text(receipt.nasabah.name || '-', 28, sigLineY + 3, { width: colWidth, align: 'center' });
      doc.text(receipt.officer_name || 'Petugas Loket', 28 + colWidth, sigLineY + 3, { width: colWidth, align: 'center' });

      doc.y = sigLineY + 18;
      doc.moveDown(0.4);

      // Footer
      doc
        .fontSize(8)
        .font('Helvetica-Oblique')
        .fillColor('#6b7280')
        .text(receipt.footer, { align: 'center' });

      doc.moveDown(0.3);
      doc
        .fontSize(7)
        .font('Helvetica')
        .fillColor('#9ca3af')
        .text(`Dicetak pada: ${receipt.printed_at}`, { align: 'center' });

      doc.end();
      } catch (err) {
        reject(err);
      }
    });
  }
}

export const receiptService = ReceiptService;
